### PR TITLE
[actpool] fix iterator issue and make code more efficient in cleanTimeout

### PR DESCRIPTION
**Pull request by Liuhaai**
### Issue 
- After deleting `i`th element in the array, the old `i+1`th element become `i`th element in the new array. So `i++` should be skipped. 
- Old deleting method would make heap invalid in certain scenarios. For instance, If `7` is removed from heap `[1, 5, 2, 6, 7, 3]`, the new heap `[1, 5, 2, 6, 3]` is invalid.
### Optimization
 - func `clock.Now()` is time-consuming. Remove it out of the loop
 - `append(q.index[:i], q.index[i+1:]...)` is O(n). Deleting the element can be achieved by swap it with the last element in the array. And reinitate the heap `heap.Init(&q.index)` in the end.

### CODE REVIEW

File: actpool/actqueue.go [hunk 1]
```diff
@@ -139,15 +139,21 @@ func (q *actQueue) FilterNonce(threshold uint64) []action.SealedEnvelope {
...
 func (q *actQueue) cleanTimeout() []action.SealedEnvelope {
-	removedFromQueue := make([]action.SealedEnvelope, 0)
-	for i := 0; i < len(q.index); i++ {
-		if q.clock.Now().After(q.index[i].deadline) {
-			// remove
+	var (
+		removedFromQueue = make([]action.SealedEnvelope, 0)
+		timeNow          = q.clock.Now()
+	)
+	for i := 0; i < len(q.index); {
```
Comment (dustinxie) [inline actpool/actqueue.go:146]: make it even faster by
```
size := len(q.index)
for i := 0; i < size; {
```

File: actpool/actqueue.go [hunk 2]
```diff
@@ -139,15 +139,21 @@ func (q *actQueue) FilterNonce(threshold uint64) []action.SealedEnvelope {
...
+		if timeNow.After(q.index[i].deadline) {
 			removedFromQueue = append(removedFromQueue, q.items[q.index[i].nonce])
 			delete(q.items, q.index[i].nonce)
-			q.index = append(q.index[:i], q.index[i+1:]...)
+			q.index[i] = q.index[len(q.index)-1]
+			q.index = q.index[:len(q.index)-1]
```
Comment (dustinxie) [inline actpool/actqueue.go:151]:
```
q.index[i] = q.index[size-1]
size--
```

File: actpool/actqueue.go [hunk 3]
```diff
@@ -139,15 +139,21 @@ func (q *actQueue) FilterNonce(threshold uint64) []action.SealedEnvelope {
...
+			continue
 		}
+		i++
 	}
+	heap.Init(&q.index)
```
Comment (dustinxie) [inline actpool/actqueue.go:156]: q.index = q.index[:size]

Comment (dustinxie) [reply actpool/actqueue.go:156]: also line 165 below can be improved:
```
removedFromQueue = q.cleanTimeout()
```

File: actpool/actqueue.go [hunk 4]
```diff
@@ -139,15 +139,23 @@ func (q *actQueue) FilterNonce(threshold uint64) []action.SealedEnvelope {
...
+		len              = len(q.index)
+	)
+	for i := 0; i < len; {
+		if timeNow.After(q.index[i].deadline) {
 			removedFromQueue = append(removedFromQueue, q.items[q.index[i].nonce])
 			delete(q.items, q.index[i].nonce)
-			q.index = append(q.index[:i], q.index[i+1:]...)
+			q.index[i] = q.index[len-1]
+			len--
+			continue
```
Comment (dustinxie) [inline actpool/actqueue.go:153]:
```
   len--
} else {
   // check next one
   i++
}
```

**Review by dustinxie — APPROVED** [review]

File: actpool/actqueue.go [hunk 5]
```diff
@@ -156,7 +164,7 @@ func (q *actQueue) UpdateQueue(nonce uint64) []action.SealedEnvelope {
 	removedFromQueue := make([]action.SealedEnvelope, 0)
 	// First remove all timed out actions
 	if q.ttl != 0 {
```
Comment (CoderZhi) [inline actpool/actqueue.go:166]: move `if q.ttl != 0` into `q.cleanTimeout`

File: actpool/actqueue.go [hunk 6]
```diff
@@ -139,15 +139,23 @@ func (q *actQueue) FilterNonce(threshold uint64) []action.SealedEnvelope {
...
 
 func (q *actQueue) cleanTimeout() []action.SealedEnvelope {
-	removedFromQueue := make([]action.SealedEnvelope, 0)
-	for i := 0; i < len(q.index); i++ {
-		if q.clock.Now().After(q.index[i].deadline) {
-			// remove
+	var (
+		removedFromQueue = make([]action.SealedEnvelope, 0)
+		timeNow          = q.clock.Now()
+		len              = len(q.index)
```
Comment (CoderZhi) [inline actpool/actqueue.go:145]: rename to `l` or `size`

**Review by CoderZhi — COMMENTED** [review]
1. what's the impact of moving q.clock.Now() out of loop
2. what's the impact of the new deletion method?
3. what's the impact of reheap?
4. what if we create a new index to store the undeleted items instead of reheap?

**Comment by Liuhaai** [comment]
> 1. what's the impact of moving q.clock.Now() out of loop
> 2. what's the impact of the new deletion method?
> 3. what's the impact of reheap?
> 4. what if we create a new index to store the undeleted items instead of reheap?

update in the description

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**